interfaces/prompting/requestprompts/requestprompts.go

```diff
@@ -0,0 +1,384 @@
+type PromptConstraints struct {
+	Path                 string   `json:"path"`
+	Permissions          []string `json:"permissions"`
+	AvailablePermissions []string `json:"available-permissions"`
+}
+
+// equals returns true if the two prompt constraints are identical.
+func (pc *PromptConstraints) equals(other *PromptConstraints) bool {
+	if pc.Path != other.Path || len(pc.Permissions) != len(other.Permissions) {
+		return false
+	}
+	// Avoid using reflect.DeepEquals to compare []string contents
+	for i := range pc.Permissions {
+		if pc.Permissions[i] != other.Permissions[i] {
+			return false
+		}
+	}
+	return true
+}
+
+// subtractPermissions removes all of the given permissions from the list of
+// permissions in the constraints.
+func (pc *PromptConstraints) subtractPermissions(permissions []string) (modified bool) {
+	newPermissions := make([]string, 0, len(pc.Permissions))
+	for _, perm := range pc.Permissions {
+		if !strutil.ListContains(permissions, perm) {
+			newPermissions = append(newPermissions, perm)
+		}
+	}
+	if len(newPermissions) != len(pc.Permissions) {
+		pc.Permissions = newPermissions
+		return true
+	}
+	return false
+}
+
+// userPromptDB maps prompt IDs to prompts for a single user.
+type userPromptDB struct {
+	ByID map[string]*Prompt
+}
+
+// PromptDB stores outstanding prompts in memory and ensures that new prompts
+// are created with a unique ID.
+type PromptDB struct {
+	perUser   map[uint32]*userPromptDB
+	maxID     uint64
+	maxIDPath string
+	mutex     sync.Mutex
+	// Function to issue a notice for a change in a prompt
+	notifyPrompt func(userID uint32, promptID string, data map[string]string) error
+}
+
+// New creates and returns a new prompt database.
+//
+// The given notifyPrompt closure should record a notice of type
+// "interfaces-requests-prompt" for the given user with the given
+// promptID as its key.
+func New(notifyPrompt func(userID uint32, promptID string, data map[string]string) error) *PromptDB {
+	pdb := PromptDB{
+		perUser:      make(map[uint32]*userPromptDB),
+		notifyPrompt: notifyPrompt,
+	}
+	// Importantly, set maxIDPath before attempting to load max ID
+	pdb.maxIDPath = filepath.Join(dirs.SnapRunDir, "request-prompt-max-id")
+	err := pdb.loadMaxID()
+	if err != nil {
+		// If cannot read max existing prompt ID, start again from 0.
+		logger.Debugf("%v; restarting at ID 0", err)
+		pdb.maxID = 0
+	}
+	return &pdb
+}
+
+// loadMaxID reads the previous ID from the file at maxIDPath and sets maxID.
+//
+// If no file exists at maxIDPath, sets maxID to be 0. If another error occurs,
+// returns it and does not set maxID.
+func (pdb *PromptDB) loadMaxID() error {
+	pdb.mutex.Lock()
+	defer pdb.mutex.Unlock()
+
+	target := pdb.maxIDPath
+	f, err := os.Open(target)
+	if os.IsNotExist(err) {
+		pdb.maxID = 0
+		return nil
+	}
+	if err != nil {
+		return fmt.Errorf("cannot read maximum prompt ID: %w", err)
+	}
+	defer f.Close()
+
+	idBuf := [16]byte{}
+	_, err = io.ReadFull(f, idBuf[:])
```

Comment:
How come we read full 16 bytes when the number is in text and not binary? Will this not err when we have less to read?

Reply:
The number is ASCII hex padded by 0s, so "0000000000000000" to "FFFFFFFFFFFFFFFF". But I prefer your suggestion of directly writing an int to a memory-mapped file.